src/aliasthis.c

```diff
@@ -21,18 +21,376 @@
 #include "mtype.h"
 #include "declaration.h"
 #include "tokens.h"
+#include "enum.h"
+#include "template.h"
+#include "arraytypes.h"
+#include "expression.h"
 
-Expression *resolveAliasThis(Scope *sc, Expression *e)
+typedef Array<const char *> Strings;
+
+/**
+ * iterateAliasThis resolves alias this subtypes for `e` and applies it to `dg`.
+ * dg should return true, if appropriate subtype has been found. Otherwise it should returns 0.
+ * Also dg can get context through secong parameter `ctx`, which was passed through fourth argument of iterateAliasThis.
+ * `dg` can return result expression through `outexpr` parameter, and if it is not NULL, it will be pushed to `ret` array.
+ * At the first stage iterateAliasThis iterates direct "alias this"-es and pushes non-null `outexpr`-es to `ret`.
+ * If `dg` for one of direct "alias this"-es returns true then `iterateAliasThis` breaks at this stage and returns true
+ * through return value and returned by `dg` expression array through `ret`.
+ * Even if true returned by first `dg` call, remaining direct "alias this"-es will be processed.
+ *
+ * If neither of the direct aliases did not return true iterateAliasThis is recursive applied to direct aliases and base
+ * classes and interfaces.
+ * If one of those `iterateAliasThis` calls returns true our `iterateAliasThis` will return true. Otherwise it will return 0.
+ *
+ * The the last argument needs for internal use and should be NULL in user call.
+ */
+
+bool iterateAliasThis(Scope *sc, Expression *e, IterateAliasThisDg dg,
+                      void *ctx, Expressions *ret, bool gagerrors, StringTable *directtypes)
+{
+    //printf("iterateAliasThis: %s\n", e->toChars());
+
+    Dsymbols *aliasThisSymbols = NULL;
+    if (e->type->toBasetype()->ty == Tstruct)
+    {
+        TypeStruct *ts = (TypeStruct *)e->type->toBasetype();
+        aliasThisSymbols = ts->sym->aliasThisSymbols;
+    }
+    else if (e->type->toBasetype()->ty == Tclass)
+    {
+        TypeClass *ts = (TypeClass *)e->type->toBasetype();
+        if (ts->sym->aliasThisSymbols)
+        {
+            aliasThisSymbols = ts->sym->aliasThisSymbols;
+        }
+    }
+    else
+    {
+        return 0;
+    }
+
+    if (!directtypes)
+    {
+        directtypes = new StringTable();
+        directtypes->_init();
+    }
+    StringValue *depth_counter = directtypes->lookup(e->type->deco, strlen(e->type->deco));
```

Comment:
Are you sure that only the one copy of the one type may be here, and type will be merged properly?
I mean that is the `type` pointers equality implies from the `deco` string equality?